File: scripts/fetch-osm-data.js
/**
 * Build script: fetch OSM data for Beijing, simplify, and output TopoJSON.
 * Run: node scripts/fetch-osm-data.js
 */

const fs = require('fs');
const path = require('path');

// We'll use native fetch (Node 18+) and minimal dependencies.
// Simplification is done with a lightweight Douglas-Peucker implementation
// to avoid pulling in the full turf bundle for a build script.

const OUTPUT_PATH = path.join(__dirname, '../public/maps/beijing-osm.json');
const BOUNDS = {
  south: 39.6359,
  north: 40.1755,
  west: 116.0396,
  east: 116.7430,
};

// ------------------------------------------------------------------
// Tiny Douglas–Peucker (operates on geo-coord arrays [lng, lat])
// ------------------------------------------------------------------
function sqDist(a, b) {
  const dx = a[0] - b[0];
  const dy = a[1] - b[1];
  return dx * dx + dy * dy;
}

function simplifyDP(points, epsilon) {
  if (points.length <= 2) return points;

  const epsilonSq = epsilon * epsilon;

  function rec(start, end, result) {
    let maxDist = -1;
    let idx = -1;
    const a = points[start];
    const b = points[end];
    for (let i = start + 1; i < end; i++) {
      // perpendicular distance squared to segment a-b
      const l2 = sqDist(a, b);
      let t = 0;
      if (l2 !== 0) {
        t = ((points[i][0] - a[0]) * (b[0] - a[0]) + (points[i][1] - a[1]) * (b[1] - a[1])) / l2;
        t = Math.max(0, Math.min(1, t));
      }
      const proj = [a[0] + t * (b[0] - a[0]), a[1] + t * (b[1] - a[1])];
      const d = sqDist(points[i], proj);
      if (d > maxDist) {
        maxDist = d;
        idx = i;
      }
    }
    if (maxDist > epsilonSq) {
      rec(start, idx, result);
      result.push(points[idx]);
      rec(idx, end, result);
    }
  }

  const result = [points[0]];
  rec(0, points.length - 1, result);
  result.push(points[points.length - 1]);
  return result;
}

function simplifyFeature(feature, epsilon) {
  const geom = feature.geometry;
  if (!geom) return feature;

  const clone = JSON.parse(JSON.stringify(feature));
  const g = clone.geometry;

  if (g.type === 'LineString') {
    g.coordinates = simplifyDP(g.coordinates, epsilon);
  } else if (g.type === 'MultiLineString') {
    g.coordinates = g.coordinates.map((line) => simplifyDP(line, epsilon));
  } else if (g.type === 'Polygon') {
    g.coordinates = g.coordinates.map((ring) => simplifyDP(ring, epsilon));
  } else if (g.type === 'MultiPolygon') {
    g.coordinates = g.coordinates.map((polygon) => polygon.map((ring) => simplifyDP(ring, epsilon)));
  }
  return clone;
}

// ------------------------------------------------------------------
// Overpass helpers
// ------------------------------------------------------------------
function buildOverpassQuery() {
  const { south, west, north, east } = BOUNDS;
  const bbox = `${south},${west},${north},${east}`;

  return `
[out:json];
(
  way["highway"~"^(motorway|trunk|primary|secondary|tertiary)$"](${bbox});
  way["waterway"="river"](${bbox});
  way["natural"="water"](${bbox});
  way["leisure"="park"](${bbox});
  way["highway"~"^(residential|living_street)$"](${bbox});
);
out body;
>;
out skel qt;
`.trim();
}

async function fetchOverpass() {
  const query = buildOverpassQuery();
  const url = 'https://overpass.openstreetmap.fr/api/interpreter';

  console.log('Fetching OSM data from Overpass API...');
  const res = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: `data=${encodeURIComponent(query)}`,
  });

  if (!res.ok) throw new Error(`Overpass API error: ${res.status} ${res.statusText}`);
  return res.json();
}

// ------------------------------------------------------------------
// Convert Overpass JSON → GeoJSON-like features
// ------------------------------------------------------------------
function parseOsmToGeoFeatures(osmJson) {
  const nodes = new Map();
  for (const el of osmJson.elements) {
    if (el.type === 'node') {
      nodes.set(el.id, [el.lon, el.lat]);
    }
  }

  const features = [];

  for (const el of osmJson.elements) {
    if (el.type !== 'way') continue;
    const coords = el.nodes.map((nid) => nodes.get(nid)).filter(Boolean);
    if (coords.length < 2) continue;

    const tags = el.tags || {};
    const highway = tags.highway || '';
    const waterway = tags.waterway || '';
    const natural = tags.natural || '';
    const leisure = tags.leisure || '';

    let category = 'road';
    let subType = 'default';

    if (highway) {
      category = 'road';
      if (['motorway', 'trunk'].includes(highway)) subType = 'motorway';
      else if (highway === 'primary') subType = 'primary';
      else if (highway === 'secondary') subType = 'secondary';
      else if (highway === 'tertiary') subType = 'tertiary';
      else if (['residential', 'living_street'].includes(highway)) subType = 'residential';
    } else if (waterway === 'river' || natural === 'water') {
      category = 'water';
      subType = waterway === 'river' ? 'river' : 'lake';
    } else if (leisure === 'park') {
      category = 'park';
      subType = 'park';
    } else {
      continue; // skip unknown
    }

    const geomType = (category === 'water' || category === 'park') ? 'Polygon' : 'LineString';
    const geometryCoords = geomType === 'Polygon' ? [coords] : coords;

    features.push({
      type: 'Feature',
      properties: { category, subType },
      geometry: {
        type: geomType,
        coordinates: geometryCoords,
      },
    });
  }

  return features;
}

// ------------------------------------------------------------------
// Manual TopoJSON encode (we only need geometry; avoids heavy lib)
// ------------------------------------------------------------------
function encodeTopoJSON(features, name) {
  const arcs = [];
  const objects = [];

  // We store everything as MultiLineString / Polygon arcs for tiny size.
  // For simplicity, encode each feature as a geometry object with arc indexes.
  function pushArc(coordList) {
    // delta-encode
    const arc = [];
    let prevLon = 0;
    let prevLat = 0;
    for (const [lon, lat] of coordList) {
      const dLon = Math.round((lon - prevLon) * 1e6);
      const dLat = Math.round((lat - prevLat) * 1e6);
      arc.push([dLon, dLat]);
      prevLon = lon;
      prevLat = lat;
    }
    const idx = arcs.length;
    arcs.push(arc);
    return idx;
  }

  const geometries = [];

  for (const f of features) {
    const geom = f.geometry;
    if (geom.type === 'LineString') {
      const arcIdx = pushArc(geom.coordinates);
      geometries.push({
        type: 'LineString',
        properties: f.properties,
        arcs: [arcIdx],
      });
    } else if (geom.type === 'Polygon') {
      const ringArcs = geom.coordinates.map((ring) => pushArc(ring));
      geometries.push({
        type: 'Polygon',
        properties: f.properties,
        arcs: [ringArcs],
      });
    } else if (geom.type === 'MultiLineString') {
      const lineArcs = geom.coordinates.map((line) => pushArc(line));
      geometries.push({
        type: 'MultiLineString',
        properties: f.properties,
        arcs: lineArcs,
      });
    } else if (geom.type === 'MultiPolygon') {
      // simplify: take first polygon only for parks/water to keep size down
      const polyArcs = geom.coordinates[0].map((ring) => pushArc(ring));
      geometries.push({
        type: 'Polygon',
        properties: f.properties,
        arcs: [polyArcs],
      });
    }
  }

  return {
    type: 'Topology',
    objects: {
      [name]: {
        type: 'GeometryCollection',
        geometries,
      },
    },
    arcs,
    bbox: [BOUNDS.west, BOUNDS.south, BOUNDS.east, BOUNDS.north],
  };
}

function decodeTopoJSON(topo) {
  const geometries = [];
  const obj = topo.objects.beijing;

  for (const g of obj.geometries) {
    const properties = g.properties;

    if (g.type === 'LineString') {
      const coords = arcToCoords(topo.arcs[g.arcs[0]]);
      geometries.push({
        type: 'Feature',
        properties,
        geometry: { type: 'LineString', coordinates: coords },
      });
    } else if (g.type === 'Polygon') {
      const rings = g.arcs[0].map((arcIdx) => arcToCoords(topo.arcs[arcIdx]));
      geometries.push({
        type: 'Feature',
        properties,
        geometry: { type: 'Polygon', coordinates: rings },
      });
    } else if (g.type === 'MultiLineString') {
      const lines = g.arcs.map((arcIdx) => arcToCoords(topo.arcs[arcIdx]));
      geometries.push({
        type: 'Feature',
        properties,
        geometry: { type: 'MultiLineString', coordinates: lines },
      });
    }
  }

  return { type: 'FeatureCollection', features: geometries };
}

function arcToCoords(arc) {
  let lon = 0;
  let lat = 0;
  const coords = [];
  for (const [dlon, dlat] of arc) {
    lon += dlon / 1e6;
    lat += dlat / 1e6;
    coords.push([lon, lat]);
  }
  return coords;
}

// ------------------------------------------------------------------
// Main
// ------------------------------------------------------------------
async function main() {
  try {
    const osmJson = await fetchOverpass();
    console.log(`Received ${osmJson.elements.length} OSM elements`);

    let features = parseOsmToGeoFeatures(osmJson);
    console.log(`Parsed ${features.length} features`);

    // Simplify per category
    const epsilons = {
      motorway: 0.0015,
      primary: 0.001,
      secondary: 0.0008,
      tertiary: 0.0006,
      default: 0.0004,
      river: 0.001,
      lake: 0.001,
      park: 0.001,
      residential: 0.0003,
    };

    features = features.map((f) => {
      const sub = f.properties.subType;
      const eps = epsilons[sub] || epsilons.default;
      return simplifyFeature(f, eps);
    });

    console.log(`Simplified features`);

    const topo = encodeTopoJSON(features, 'beijing');

    // verify decode works
    const decoded = decodeTopoJSON(topo);
    console.log(`Decoded back to ${decoded.features.length} features for sanity check`);

    const jsonStr = JSON.stringify(topo);
    fs.mkdirSync(path.dirname(OUTPUT_PATH), { recursive: true });
    fs.writeFileSync(OUTPUT_PATH, jsonStr, 'utf-8');

    const sizeKB = (jsonStr.length / 1024).toFixed(1);
    console.log(`Wrote ${OUTPUT_PATH} (${sizeKB} KB)`);
  } catch (err) {
    console.error('Failed to fetch OSM data:', err.message);
    process.exit(1);
  }
}

main();

File: scripts/fetch-osm-data.py
#!/usr/bin/env python3
"""
Build script: fetch OSM data for Beijing using osmnx, aggressively simplify,
pre-project to SVG coordinates, and output a compact JSON.

Strategy: only keep motorways + trunks + largest water bodies + largest parks.
Target: < 200 KB.

Run: python3 scripts/fetch-osm-data.py
"""

import json
import os

import osmnx as ox
from shapely import simplify
from shapely.geometry import mapping

OUTPUT_PATH = os.path.join(os.path.dirname(__file__), "../public/maps/beijing-osm.json")

BOUNDS = {
    "south": 39.6359,
    "north": 40.1755,
    "west": 116.0396,
    "east": 116.7430,
}

W, E, S, N = BOUNDS["west"], BOUNDS["east"], BOUNDS["south"], BOUNDS["north"]
DW = E - W
DH = N - S


def project(lon, lat):
    x = ((lon - W) / DW) * 1000
    y = ((N - lat) / DH) * 800
    return (round(x, 1), round(y, 1))


def simplify_ring(ring, epsilon):
    if len(ring) <= 2:
        return ring

    def sq_dist(a, b):
        dx = a[0] - b[0]
        dy = a[1] - b[1]
        return dx * dx + dy * dy

    def rec(start, end, result):
        max_dist = -1
        idx = -1
        a = ring[start]
        b = ring[end]
        l2 = sq_dist(a, b)
        for i in range(start + 1, end):
            if l2 == 0:
                t = 0
            else:
                t = max(
                    0.0,
                    min(
                        1.0,
                        (
                            (ring[i][0] - a[0]) * (b[0] - a[0])
                            + (ring[i][1] - a[1]) * (b[1] - a[1])
                        )
                        / l2,
                    ),
                )
            proj = (a[0] + t * (b[0] - a[0]), a[1] + t * (b[1] - a[1]))
            d = sq_dist(ring[i], proj)
            if d > max_dist:
                max_dist = d
                idx = i
        if max_dist > epsilon * epsilon:
            rec(start, idx, result)
            result.append(ring[idx])
            rec(idx, end, result)

    result = [ring[0]]
    rec(0, len(ring) - 1, result)
    result.append(ring[-1])
    return result


def fetch_roads():
    print("Fetching motorways + trunks only...")
    bbox = (BOUNDS["west"], BOUNDS["south"], BOUNDS["east"], BOUNDS["north"])

    # Only motorways and trunks to keep data tiny
    custom_filter = '["highway"~"^(motorway|trunk)$"]'
    G = ox.graph_from_bbox(
        bbox,
        network_type="drive",
        simplify=True,
        retain_all=True,
        custom_filter=custom_filter,
    )

    gdf_nodes, gdf_edges = ox.graph_to_gdfs(G)
    print(f"Road edges: {len(gdf_edges)}")

    roads = []
    seen = set()
    for _, row in gdf_edges.iterrows():
        geom = row.geometry
        if geom is None or geom.is_empty:
            continue

        coords = list(geom.coords)
        if len(coords) < 2:
            continue

        svg_coords = [project(lon, lat) for lon, lat in coords]
        svg_coords = simplify_ring(svg_coords, 3.0)  # aggressive simplify in px
        if len(svg_coords) < 2:
            continue

        # dedup
        key = tuple((round(x, 1), round(y, 1)) for x, y in svg_coords)
        rkey = tuple(reversed(key))
        if key in seen or rkey in seen:
            continue
        seen.add(key)

        roads.append({"t": "motorway", "c": [svg_coords]})

    return roads


def fetch_features(tags, category, sub_type, geo_eps, min_area, top_n):
    print(f"Fetching {category} ({sub_type})...")
    bbox = (BOUNDS["west"], BOUNDS["south"], BOUNDS["east"], BOUNDS["north"])
    try:
        gdf = ox.features_from_bbox(bbox, tags=tags)
    except Exception as e:
        print(f"  Failed: {e}")
        return []

    print(f"  Found {len(gdf)} features")

    items = []
    for _, row in gdf.iterrows():
        geom = row.geometry
        if geom is None or geom.is_empty:
            continue

        # Use area for polygons, length for linestrings
        try:
            if geom.geom_type in ("Polygon", "MultiPolygon"):
                metric = geom.area
            else:
                metric = geom.length
        except Exception:
            metric = 0

        if metric < min_area:
            continue

        try:
            geom = simplify(geom, tolerance=geo_eps)
        except Exception:
            pass

        svg_rings = []
        if geom.geom_type == "Polygon":
            rings = [list(geom.exterior.coords)] + [list(i.coords) for i in geom.interiors]
            for ring in rings:
                svg_ring = [project(lon, lat) for lon, lat in ring]
                svg_ring = simplify_ring(svg_ring, 2.0)
                if len(svg_ring) >= 3:
                    svg_rings.append(svg_ring)
        elif geom.geom_type == "MultiPolygon":
            for poly in geom.geoms:
                rings = [list(poly.exterior.coords)] + [list(i.coords) for i in poly.interiors]
                for ring in rings:
                    svg_ring = [project(lon, lat) for lon, lat in ring]
                    svg_ring = simplify_ring(svg_ring, 2.0)
                    if len(svg_ring) >= 3:
                        svg_rings.append(svg_ring)
        elif geom.geom_type == "LineString":
            svg_line = [project(lon, lat) for lon, lat in geom.coords]
            svg_line = simplify_ring(svg_line, 2.0)
            if len(svg_line) >= 2:
                svg_rings.append(svg_line)
        elif geom.geom_type == "MultiLineString":
            for line in geom.geoms:
                svg_line = [project(lon, lat) for lon, lat in line.coords]
                svg_line = simplify_ring(svg_line, 2.0)
                if len(svg_line) >= 2:
                    svg_rings.append(svg_line)

        if svg_rings:
            items.append({"metric": metric, "rings": svg_rings})

    # keep top N
    items.sort(key=lambda x: x["metric"], reverse=True)
    items = items[:top_n]
    print(f"  Kept top {len(items)}")

    return [{"t": sub_type, "c": item["rings"]} for item in items]


def main():
    print("Fetching Beijing OSM data (aggressive minimal)...")

    roads = fetch_roads()
    # rivers are linestrings -> use length threshold (in degrees)
    rivers = fetch_features({"waterway": "river"}, "water", "river", 0.005, min_area=0.003, top_n=20)
    # lakes/parks are polygons -> use area threshold (in square degrees); 0.00005 ~ 0.5 km2 at this latitude
    lakes = fetch_features({"natural": "water"}, "water", "lake", 0.005, min_area=0.00005, top_n=20)
    parks = fetch_features({"leisure": "park"}, "park", "park", 0.005, min_area=0.00005, top_n=20)

    data = {
        "type": "beijing-map",
        "bounds": BOUNDS,
        "features": roads + rivers + lakes + parks,
    }

    os.makedirs(os.path.dirname(OUTPUT_PATH), exist_ok=True)
    with open(OUTPUT_PATH, "w", encoding="utf-8") as f:
        json.dump(data, f, separators=(",", ":"))

    size_kb = os.path.getsize(OUTPUT_PATH) / 1024
    print(f"Wrote {OUTPUT_PATH} ({size_kb:.1f} KB)")
    print(f"  Roads: {len(roads)}")
    print(f"  Rivers: {len(rivers)}")
    print(f"  Lakes: {len(lakes)}")
    print(f"  Parks: {len(parks)}")


if __name__ == "__main__":
    main()
